### PR TITLE
Feature/frontend map timeline update

### PR DESCRIPTION
**Pull request by cadw**
I have finished the map and the timeline. Everything works. There is, however, one little bug. Sometimes, the database is not at the right time, and the wrong timepoint is used. We can fix this by removing the timepoints and add a startCue and duration attribute. I think it is better to fix this in another branch. I have also removed the grid image. Now the grid is drawn using FabricJS, which gives a better look.

### CODE REVIEW

**Comment by ChrisBerg** [comment]
Please resolve merge conflicts

File: src/main/webapp/app/scripting/scriptingview.controller.js [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+            for (var j = 0; j < 4 - year.toString().length; ++j)
+                current += '0';
+
+            current += year.toString();
+
+            return current;
+        }
+
+        function loadCues () {
```
Comment (ChrisBerg) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: Incredibly long function

File: src/main/webapp/app/scripting/scriptingview.state.js [hunk 2]
```diff
@@ -7,21 +7,17 @@
...
-            url: '/timeline',
+            url: '/scripting',
             data: {
                 authorities: ['ROLE_USER']
             },
             views: {
                 'content@': {
-                    templateUrl: 'app/timeline/timeline.html',
-                    controller: 'TimelineController',
+                    templateUrl: 'app/scripting/scriptingview.html',
```
Comment (ChrisBerg) [inline src/main/webapp/app/scripting/scriptingview.state.js]: ScriptingView is kind of a duplication, scripting would suffice i think

File: src/main/webapp/app/scripting/scriptingview.controller.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+     * @param $state the state of the map
+     * @param Camera the camera entity
+     * @param AlertService the alertservice
+     * @constructor
+     */
+    function ScriptingviewController ($scope, $state, Camera, Player, CameraAction, Script, TimePoint, Cue, AlertService) {
+        var vm = this;
+        var grid = 15;
+
+        vm.loadCameras = loadCameras;
```
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: You can simply call `vm.cameras = Camera.query();`.

File: src/main/webapp/app/scripting/scriptingview.controller.js [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        // vm.loadTimePoints();
+
+        function loadTimePoints () {
+            TimePoint.query({
+
+            }, onSuccess, onError);
+
+            function onSuccess(data, headers) {
+                vm.timePoints = data;
+                vm.queryCount = vm.totalItems;
```
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: `vm.totalItems` isn't defined anywhere? Do you actually use `vm.queryCount` somewhere?

File: src/main/webapp/app/scripting/scriptingview.controller.js [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+                    'timeAxis' : {scale: 'year', step: 1},
+                    'min': '0000-01-01',
+                    'zoomMin': 63072000000,
+                    'zoomMax': 700000000000
+                };
+
+                // Create a Timeline
+                var timeline = new vis.Timeline(container, items, options);
+
+                timeline.on('click', function (properties) {
```
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: Incredibly long and error-prone. Why does it refresh the whole page when you click on it?

File: src/main/webapp/app/scripting/scriptingview.controller.js (hunk 3 above)
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: Same thing applies for querying all other things.

**Comment by ChrisBerg** [comment]
Nice work on the scripting page, keep up the good work!

A small remark on the Cue creation:
When i click on the time bar once (in order to add a cue), all cameras pile up at the top left corner.
When i click on the time bar a second time all cameras move to the right position again. On top of that, a cue is created at the time point where i clicked the first time, which kind of makes it feel like it is broken. Is there a fast way to fix this?

**Comment by cadw** [comment]
This is caused by JavaScript Asynchronous calls, where not all objects are fully loaded at startup. This is usually caused by develop mode and should not occur when launched.

File: src/main/webapp/app/scripting/scriptingview.controller.js (hunk 5 above)
Comment (cadw) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: It has to reload all cues in order to draw them in the timeline, doing that in the same session does not load the most recent camera.

File: src/main/webapp/app/scripting/scriptingview.controller.js (hunk 3 above)
Comment (cadw) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: No you cannot

**Comment by ChrisBerg** [comment]
Makes sense, i see that there is an issue too.

**Comment by hungnguyen94** [comment]
That controller method is way too long and buggy. You should also refrain from using `document.getElementById`. We use AngularJS, so you should take advantage of the tools it provides. 
Look into scoping and directives.

File: src/main/webapp/app/scripting/scriptingview.controller.js (hunk 5 above)
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: What? 
Why reload all the cues? You could just add the new Cue you created to the timeline.

**Comment by hungnguyen94** [comment]
You can add and remove Player and Camera objects on the map now. You can also move them around and they will get updated automatically.   
For some reason Travis is not running?

**Comment by hungnguyen94** [comment]
Fixed tests, this branch can finally be merged \o/

File: src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java [hunk 6]
```diff
@@ -77,46 +76,48 @@ public Map findOne(Long id) {
...
+//    @Transactional
+//    public Optional<Map> addPlayer(Long mapId, Long playerId) {
+//        Map map = findOne(mapId);
+//        return Optional.ofNullable(playerService.findOne(playerId))
+//                .map(player -> {
+//                    log.debug("Request to add player {} to map {}", mapId, playerId);
+//                    map.addPlayer(player);
+//                    return mapRepository.save(map);
+//                });
+//    }
```
Comment (rubenwiersma) [inline src/main/java/nl/tudelft/thefirstorder/service/impl/MapServiceImpl.java]: Commented, redundant code

File: src/main/java/nl/tudelft/thefirstorder/service/MapService.java [hunk 7]
```diff
@@ -43,21 +42,21 @@
...
+//     */
+//    Optional<Map> addCamera(Long mapId, Long cameraId);
+//
+//    /**
+//     * Adds a player to the Map.
+//     * @param mapId Id of the Map
+//     * @param playerId Id of the Player
+//     * @return The updated map
+//     */
+//    Optional<Map> addPlayer(Long mapId, Long playerId);
```
Comment (rubenwiersma) [inline src/main/java/nl/tudelft/thefirstorder/service/MapService.java]: Commented code

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            restrict: 'EA',
+            link: link,
+            // templateUrl: 'app/map/fabric-map-view-template.html',
+            // scope: {
+            //     map: '='
+            //     // selected: '='
+            //     // onSelect: '&', 
+            //     // onClick: '&'
+            // }, 
+            // controller: 'FabricMapController',
```
Comment (rubenwiersma) [inline src/main/webapp/app/map/fabric-map.directive.js]: Commented

File: src/main/webapp/app/map/fabric-map.directive.js [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+                drawable.lockScalingY = true;
+                // drawable.hasControls = false;
+                var position = getAbsolutePosition(drawable.x, drawable.y);
+                drawable.left = position.x;
+                drawable.top = position.y;
+                return drawable;
+            }
+
+            function transformCamera(camera) {
+                var cam = new fabric.Path('m96.6,26.8h-86.1c-2.2,0-4.1,1.8-4.1,4.1v67.2c0,2.2 1.8,4.1 4.1,4.1h86.1c2.2,0 4.1-1.8 4.1-4.1v-19.4l14.9,14.9c0.8,0.8 1.8,1.2 2.9,1.2 0.5,0 1.1-0.1 1.6-0.3 1.5-0.6 2.5-2.1 2.5-3.8v-52.5c0-1.6-1-3.1-2.5-3.8-1.5-0.6-3.3-0.3-4.4,0.9l-14.9,14.9v-19.3c-0.1-2.3-1.9-4.1-4.2-4.1zm-4.1,33.3v8.8 25.2h-78v-59.2h78v25.2zm21.9-12v32.9l-13.7-13.7v-5.4l13.7-13.8z');
```
Comment (rubenwiersma) [inline src/main/webapp/app/map/fabric-map.directive.js]: Could this not be loaded from somewhere?

File: src/main/webapp/app/map/map-editor.controller.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        console.log('MapViewController this: ', this);
+        var vm = this;
+        vm.selected = null;
+        
+        vm.setSelected = setSelected;
+        vm.update = getMapEntities;
+        getMapEntities();
+
+
+        // $scope.$watch('vm.modified', onModify);
```
Comment (rubenwiersma) [inline src/main/webapp/app/map/map-editor.controller.js]: Commented

File: src/main/webapp/app/map/map-editor.controller.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        
+        function getMapEntities() {
+            Map.getDTO({id: 1}, function (result) {
+                console.log('result is: ', result);
+                // vm.eenwillekeurigemap = result;
```
Comment (rubenwiersma) [inline src/main/webapp/app/map/map-editor.controller.js]: Commented

File: src/main/webapp/app/map/map-table-view-template.html [hunk 12]
```diff
@@ -0,0 +1,94 @@
+<!--<button type="button" class="btn btn-raised" ng-click="vm.isCamerasCollapsed = !vm.isCamerasCollapsed">Toggle cameras collapse</button>-->
+<!--<button type="button" class="btn btn-raised" ng-click="vm.isPlayersCollapsed = !vm.isPlayersCollapsed">Toggle players collapse</button>-->
+<button type="button" class="btn btn-raised" ng-click="vm.update()">Update</button>
+
+
+<!--<div class="row">-->
+<!--<div class="col-md-8">-->
+<!--<canvas id="fabricCanvas"></canvas>-->
+<!--</div>-->
+<!--<div class="col-md-4">-->
```
Comment (rubenwiersma) [inline src/main/webapp/app/map/map-table-view-template.html]: Commented

File: src/main/webapp/app/scripting/scriptingview.controller.js [hunk 13]
```diff
@@ -24,6 +24,27 @@
         vm.loadPlayers = loadPlayers;
         vm.loadCues = loadCues;
         vm.loadCues();
+        vm.loadCameraActions = loadCameraActions;
+        vm.loadCameraActions();
+        vm.loadScripts = loadScripts;
+        vm.loadScripts();
+        vm.loadTimePoints = loadTimePoints;
+        // vm.loadTimePoints();
```
Comment (rubenwiersma) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: Commented

File: src/main/webapp/app/map/fabric-map.directive.js (hunk 9 above)
Comment (hungnguyen94) [inline src/main/webapp/app/map/fabric-map.directive.js]: You could load the file which will return a callback with the object, but I wasn't exactly sure how to do that. Perhaps someone else can look into that.

File: src/main/webapp/app/scripting/scriptingview.controller.js (hunk 13 above)
Comment (hungnguyen94) [inline src/main/webapp/app/scripting/scriptingview.controller.js]: That's not my code though.

**Comment by hungnguyen94** [comment]
I removed all the commented code. Please check if I haven't missed anything.

**Comment by rubenwiersma** [comment]
Okay, it looks mergeable!

**Pull request merged**